### ISSUE TITLE
logger consume all thread resources

### ISSUE THREAD

**Issue by Stackle** [dwbutler/logstash-logger#67]
I just found a problem when I using logger with LogStash connect through TCP and I'm using ElasticSearch as LogStash's storage.

In my website, every times user open website, I will start broadcast event asynchronously (using wisper and wisper-celluloid gem) to log user access into LogStash. 

Under the hood, every times when you broadcast event, wisper will start new celluloid actor which will spawn new thread and under that actor thread it will try to send log to LogStash.

I don't want to block website request to send the log first, that's why I use wisper and celluloid.

On my server, Somethings accidentally stop ElasticSearch service but LogStash is still running then, 30 minutes later, my server is down. It eat up all thread resources in my server.

I try to find which thread is not getting killed and I found that it is logger thread. this is backtrace of that log:
```
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:185:in `lock'
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:185:in `mon_enter'
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:209:in `mon_synchronize'
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/logger.rb:559:in `write'
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/logger.rb:378:in `add'
/home/myweb/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/logger.rb:434:in `info'
/home/myweb/web/app/wisper_subscribers/Logger.rb:114:in `log_user_access'
```
Thousands of threads are hanging at `lock` method.

After I found that cause and proved that, by not start ElasticSearch service, is a root of problem then, I just start ElasticSearch and then every thread that locking is starting to finish their job. So, if LogStash and ElasticSearch is running normally, everything is fine.

**Comment by dwbutler** [comment]
Hi,

Here is my hypothesis about what is going on here.

You are running a logstash listener which is pushing incoming log messages into Elasticsearch. If Elasticsearch goes down, the logstash listener simply buffers the messages. Eventually, it runs out of buffer space and stops responding to further requests.

You mentioned that you are communicating with the logstash listener over TCP. Each message sent over TCP waits for a response. If the logstash listener stops responding, the TCP connection essentially gets stuck and cannot process any further messages.

Normally this isn't such a big deal because LogStashLogger buffers messages on its side. However, this has a limit. Eventually you will run out of buffer space, and then further attempts to write to the logger will block. For more explanation and tuning advice, see https://github.com/dwbutler/logstash-logger#buffering--automatic-retries

30 minutes is too long for a critical piece of infrastructure to be down. Cascading side effects are inevitable. You could probably tolerate 30 seconds of downtime without any issues.

If this is an ongoing issue for you, you may want to try communicating with the logstash listener over UDP instead of TCP. UDP communication doesn't wait for a response. This will keep your web app responsive, at the cost of dropping log messages.

Hope that helps.

**Comment by dwbutler** [comment]
Another possible solution to your problem is to run a logstash listener locally on each web server, and communicate with the listener over a Unix socket. Then Elasticsearch can be down as long as you like, and log messages will be buffered in the socket file while the listener waits for Elasticsearch to come back up.

**Comment by dwbutler** [comment]
This should be fixed by #81. This fixes the default behavior of connectable devices so that they are always non-blocking, even when there are connection failures. Messages are re-queued when a connection failure occurs. Messages are discarded when the buffer is full.

**Comment by dwbutler** [comment]
This behavior should be fixed in 0.16.0.
